### ISSUE TITLE
Comment on newer python

### ISSUE THREAD

**Issue by jzheng25** [PENGLU-WashU/IMC_Denoise#37]
Dear Team!

Firstly, thank you for developing this wonderful tool! I have recent upgrade with newer GPU (A4000), which result in older CUDA not working well.
After a couple test and run, I would like to share a workable environment with Python 3.8
- My working environment is in **WIN11 with CUDA11.8 and cuDNN8.6 installed in windows**

### CREATE AN ENVIRONMENT IN PYTHON 3.8 ###
conda create -n imc_denoise python=3.8 -y

### ACTIVATE CONDA ENVIRONMENT ###
conda activate imc_denoise 

### INSTALL PACKAGE ###
    # use tensorflow 2.6 and no need for keras anymore
    # need to modify setup.py to avoid auto-versioning
pip install tensorflow==2.6.0 
conda install -c conda-forge cudatoolkit=11.2 cudnn=8.1

### INSTALL IMC_Denoise ###
cd IMC_Denoise # in setup.py set tensorflow ==2.6.0 and remove keras
pip install -e .

### INSTALL AND FIX JUPYTER ###
conda install jupyter
pip uninstall charset_normalizer -y
pip install charset_normalizer==2.1.1
pip uninstall requests -y
pip install requests

### FIX TF dependency ### (may add it to setup.py?)  
pip install protobuf==3.20.*

### MODIFY CODE ###
In server place, change keras to tensorflow.keras

**Comment by jzheng25** [comment]
I found most of the issues was address in TF26. May also add that to front page to compatible with modern GPU

**Comment by dr-michael-haley** [comment]
This worked for me after some tweaking - I also needed to play around with some other package versions (e.g scipy and numpy), eventually getting this combination working on an RTX A6000 GPU. 

The original code is inconsistent where it uses either `import keras`, or `import tensorflow.keras`, which is a bit confusing, so you may need to do some trial and error, and change them to `tensorflow.keras`

`conda list`...
**Name                     Version             Build               Channel**
absl-py                    0.15.0              pypi_0              pypi
aom                        3.5.0               h63175ca_0          conda-forge
astunparse                 1.6.3               pypi_0              pypi
blas                       1.0                 mkl                 conda-forge
blosc                      1.21.5              hdccc3a2_0          conda-forge
brotlipy                   0.7.0               py38h91455d4_1006   conda-forge
bzip2                      1.0.8               h2466b09_7          conda-forge
c-blosc2                   2.12.0              h183a6f4_0          conda-forge
ca-certificates            2025.6.15           h4c7d964_0          conda-forge
cachetools                 5.5.2               pypi_0              pypi
certifi                    2025.6.15           pypi_0              pypi
cffi                       1.17.0              py38h4cb3324_0      conda-forge
cfitsio                    4.2.0               h9ebe7e4_0          conda-forge
charls                     2.4.2               h1537add_0          conda-forge
charset-normalizer         3.4.2               pypi_0              pypi
clang                      5.0                 pypi_0              pypi
cudatoolkit                11.2.2              h7d7167e_13         conda-forge
cudnn                      8.1.0.77            h3e0f4f4_0          conda-forge
cycler                     0.12.1              pyhd8ed1ab_0        conda-forge
dav1d                      1.2.1               hcfcfb64_0          conda-forge
flatbuffers                1.12                pypi_0              pypi
freetype                   2.13.3              h57928b3_1          conda-forge
gast                       0.4.0               pypi_0              pypi
gettext                    0.22.5              h5728263_3          conda-forge
gettext-tools              0.22.5              h5a7288d_3          conda-forge
giflib                     5.2.2               h64bf75a_0          conda-forge
glib                       2.84.2              he8f994d_0          conda-forge
glib-tools                 2.84.2              h4394cf3_0          conda-forge
google-auth                2.40.3              pypi_0              pypi
google-auth-oauthlib       1.0.0               pypi_0              pypi
google-pasta               0.2.0               pypi_0              pypi
grpcio                     1.70.0              pypi_0              pypi
gst-plugins-base           1.22.9              h001b923_1          conda-forge
gstreamer                  1.22.9              hb4038d2_1          conda-forge
h5py                       3.1.0               pypi_0              pypi
icc_rt                     2022.1.0            h6049295_2
icu                        73.2                h63175ca_0          conda-forge
idna                       3.10                pypi_0              pypi
imagecodecs                2023.1.23           py38h29c013e_2      conda-forge
imc-denoise                0.0.0               dev_0               <develop>
importlib-metadata         8.5.0               pypi_0              pypi
intel-openmp               2025.1.0            h57928b3_980        conda-forge
joblib                     1.4.2               pyhd8ed1ab_0        conda-forge
jxrlib                     1.1                 hcfcfb64_3          conda-forge
keras                      2.6.0               pypi_0              pypi
keras-preprocessing        1.1.2               pypi_0              pypi
kiwisolver                 1.4.5               py38hb1fd069_1      conda-forge
krb5                       1.21.3              hdf4eb48_0          conda-forge
lcms2                      2.15                h3e3b177_1          conda-forge
lerc                       4.0.0               h6470a55_1          conda-forge
libaec                     1.1.4               h20038f6_0          conda-forge
libasprintf                0.22.5              h5728263_3          conda-forge
libasprintf-devel          0.22.5              h5728263_3          conda-forge
libavif                    0.11.1              ha7b97ba_2          conda-forge
libblas                    3.9.0               8_mkl               conda-forge
libbrotlicommon            1.0.9               hcfcfb64_9          conda-forge
libbrotlidec               1.0.9               hcfcfb64_9          conda-forge
libbrotlienc               1.0.9               hcfcfb64_9          conda-forge
libcblas                   3.9.0               8_mkl               conda-forge
libclang                   15.0.7              default_h3a3e6c3_5  conda-forge
libclang13                 15.0.7              default_hf64faad_5  conda-forge
libcurl                    8.14.1              h88aaa65_0          conda-forge
libdeflate                 1.18                hcfcfb64_0          conda-forge
libffi                     3.4.6               h537db12_1          conda-forge
libfreetype                2.13.3              h57928b3_1          conda-forge
libfreetype6               2.13.3              h0b5ce68_1          conda-forge
libgettextpo               0.22.5              h5728263_3          conda-forge
libgettextpo-devel         0.22.5              h5728263_3          conda-forge
libglib                    2.84.2              hbc94333_0          conda-forge
libiconv                   1.18                h135ad9c_1          conda-forge
libintl                    0.22.5              h5728263_3          conda-forge
libintl-devel              0.22.5              h5728263_3          conda-forge
libjpeg-turbo              2.1.5.1             hcfcfb64_1          conda-forge
liblapack                  3.9.0               8_mkl               conda-forge
liblzma                    5.8.1               h2466b09_2          conda-forge
liblzma-devel              5.8.1               h2466b09_2          conda-forge
libogg                     1.3.5               h2466b09_1          conda-forge
libpng                     1.6.49              h7a4582a_0          conda-forge
libsqlite                  3.50.1              h67fdade_0          conda-forge
libssh2                    1.11.1              h9aa295b_0          conda-forge
libtiff                    4.5.1               h6c8260b_1          conda-forge
libvorbis                  1.3.7               h0e60522_0          conda-forge
libwebp-base               1.5.0               h3b0e114_0          conda-forge
libxcb                     1.15                hcd874cb_0          conda-forge
libxml2                    2.13.8              h442d1da_0          conda-forge
libzlib                    1.3.1               h2466b09_2          conda-forge
libzopfli                  1.0.3               h0e60522_0          conda-forge
lz4-c                      1.9.4               hcfcfb64_0          conda-forge
m2w64-gcc-libgfortran      5.3.0               6                   conda-forge
m2w64-gcc-libs             5.3.0               7                   conda-forge
m2w64-gcc-libs-core        5.3.0               7                   conda-forge
m2w64-gmp                  6.1.0               2                   conda-forge
m2w64-libwinpthread-git    5.0.0.4634.697f757  2                   conda-forge
markdown                   3.7                 pypi_0              pypi
markupsafe                 2.1.5               pypi_0              pypi
matplotlib                 3.3.4               py38haa244fe_0      conda-forge
matplotlib-base            3.3.4               py38h34ddff4_0      conda-forge
mkl                        2020.4              hb70f87d_311        conda-forge
mkl-service                2.3.0               py38h1e8a9f7_2      conda-forge
msys2-conda-epoch          20160418            1                   conda-forge
numpy                      1.21.0              pypi_0              pypi
oauthlib                   3.3.0               pypi_0              pypi
openjpeg                   2.5.0               ha2aaf27_2          conda-forge
openssl                    3.5.0               ha4e3fda_1          conda-forge
opt-einsum                 3.3.0               pypi_0              pypi
packaging                  25.0                pyh29332c3_1        conda-forge
pandas                     1.1.5               py38hf11a4ad_0
pcre2                      10.45               h99c9b8b_0          conda-forge
pillow                     10.0.0              py38ha7eb54a_0      conda-forge
pip                        24.3.1              pyh8b19718_0        conda-forge
ply                        3.11                pyhd8ed1ab_2        conda-forge
protobuf                   3.20.3              pypi_0              pypi
pthread-stubs              0.4                 hcd874cb_1001       conda-forge
pyasn1                     0.6.1               pypi_0              pypi
pyasn1-modules             0.4.2               pypi_0              pypi
pycparser                  2.22                pyhd8ed1ab_0        conda-forge
pyparsing                  3.1.4               pyhd8ed1ab_0        conda-forge
pyqt                       5.15.9              py38hd6c051e_5      conda-forge
pyqt5-sip                  12.12.2             py38hd3f51b4_5      conda-forge
python                     3.8.20              hfaddaf0_2_cpython  conda-forge
python-dateutil            2.9.0               pyhd8ed1ab_0        conda-forge
python_abi                 3.8                 7_cp38              conda-forge
pytz                       2024.2              pyhd8ed1ab_0        conda-forge
pyyaml                     6.0.2               py38h4cb3324_0      conda-forge
qt-main                    5.15.8              he5a7383_16         conda-forge
requests                   2.32.4              pypi_0              pypi
requests-oauthlib          2.0.0               pypi_0              pypi
rsa                        4.9.1               pypi_0              pypi
scikit-learn               0.24.2              py38h8224a6f_1      conda-forge
scipy                      1.4.1               py38h9439919_0
setuptools                 75.3.0              pyhd8ed1ab_0        conda-forge
sip                        6.7.12              py38hd3f51b4_0      conda-forge
six                        1.15.0              pypi_0              pypi
snappy                     1.1.10              hfb803bf_1          conda-forge
spatialbiologytoolkit      0.1                 dev_0               <develop>
tensorboard                2.14.0              pypi_0              pypi
tensorboard-data-server    0.7.2               pypi_0              pypi
tensorflow                 2.6.0               pypi_0              pypi
tensorflow-estimator       2.15.0              pypi_0              pypi
termcolor                  1.1.0               pypi_0              pypi
threadpoolctl              3.5.0               pyhc1e730c_0        conda-forge
tifffile                   2023.7.10           pyhd8ed1ab_0        conda-forge
tk                         8.6.13              h2c6b04d_2          conda-forge
toml                       0.10.2              pyhd8ed1ab_0        conda-forge
tomli                      2.0.2               pyhd8ed1ab_0        conda-forge
tornado                    6.4.1               py38h4cb3324_0      conda-forge
typing-extensions          3.7.4.3             pypi_0              pypi
ucrt                       10.0.22621.0        h57928b3_1          conda-forge
urllib3                    2.2.3               pypi_0              pypi
vc                         14.3                h2b53caa_26         conda-forge
vc14_runtime               14.42.34438         hfd919c2_26         conda-forge
vs2015_runtime             14.42.34438         h7142326_26         conda-forge
werkzeug                   3.0.6               pypi_0              pypi
wheel                      0.45.1              pyhd8ed1ab_0        conda-forge
wrapt                      1.12.1              pypi_0              pypi
xorg-libxau                1.0.11              hcd874cb_0          conda-forge
xorg-libxdmcp              1.1.3               hcd874cb_0          conda-forge
xz                         5.8.1               h208afaa_2          conda-forge
xz-tools                   5.8.1               h2466b09_2          conda-forge
yaml                       0.2.5               h8ffe710_2          conda-forge
zfp                        1.0.1               he0c23c2_2          conda-forge
zipp                       3.20.2              pypi_0              pypi
zlib-ng                    2.0.7               hcfcfb64_0          conda-forge
zstd                       1.5.7               hbeecb71_2          conda-forge
